apps/client/src/app/modules/freecompany-workshops/+state/freecompany-workshop.facade.ts

```diff
@@ -0,0 +1,325 @@
+      [35, 36, 37, 38], // rigging
+      [40, 41, 42, 43], // forecastle
+      [45, 46, 47, 48] // aftcastle
+    ];
+    const submarineSlots = [
+      [0, 1, 2, 3], // hull
+      [5, 6, 7, 8], // stern
+      [10, 11, 12, 13], // bow
+      [15, 16, 17, 18] // bridge
+    ];
+
+    if (itemInfo.containerId === 25003) {
+      partCondition.type = VesselType.AIRSHIP;
+      for (let i = 0; i < 4; i++) {
+        if (airshipSlots[i].includes(itemInfo.slot)) {
+          partCondition.vesselSlot = i;
+          partCondition.partSlot = airshipSlots[i].indexOf(itemInfo.slot);
+          break;
+        }
+      }
+    } else if (itemInfo.containerId === 25004) {
+      partCondition.type = VesselType.SUBMARINE;
+      for (let i = 0; i < 4; i++) {
+        if (submarineSlots[i].includes(itemInfo.slot)) {
+          partCondition.vesselSlot = i;
+          partCondition.partSlot = submarineSlots[i].indexOf(itemInfo.slot);
+          break;
+        }
+      }
+    }
+
+    return partCondition.type !== undefined ? partCondition : null;
+  }
+
+  public updateVesselPartCondition(packet: ItemInfo | UpdateInventorySlot): void {
+    const partUpdate = this.getVesselPartCondition(packet);
+    this.store.dispatch(FreecompanyWorkshopActions.updateVesselPart({ vesselPartUpdate: partUpdate }));
+  }
+
+  public updateVesselTimers(data): void {
+    this.store.dispatch(FreecompanyWorkshopActions.updateVesselTimers({
+      vesselTimersUpdate: {
+        type: data.type,
+        timers: data.timers
+      }
+    }));
+  }
+
+  public getRemainingTime(unixTimestamp) {
+    return unixTimestamp - Math.floor(Date.now() / 1000);
+  }
+
+  public getSubmarineBuild(submarine: Submarine) {
+    return {
+      hull: this.lazyData.data.submarineParts[submarine.parts.hull.partId],
+      stern: this.lazyData.data.submarineParts[submarine.parts.stern.partId],
+      bow: this.lazyData.data.submarineParts[submarine.parts.bow.partId],
+      bridge: this.lazyData.data.submarineParts[submarine.parts.bridge.partId]
+    };
+  }
+
+  public getSubmarineStats(submarine: Submarine): VesselStats {
+    const hull = this.lazyData.data.submarineParts[submarine.parts.hull.partId];
+    const stern = this.lazyData.data.submarineParts[submarine.parts.stern.partId];
+    const bow = this.lazyData.data.submarineParts[submarine.parts.bow.partId];
+    const bridge = this.lazyData.data.submarineParts[submarine.parts.bridge.partId];
+    return {
+      surveillance: this.sumStat('surveillance', submarine.rank, hull, stern, bow, bridge),
+      retrieval: this.sumStat('retrieval', submarine.rank, hull, stern, bow, bridge),
+      speed: this.sumStat('speed', submarine.rank, hull, stern, bow, bridge),
+      range: this.sumStat('range', submarine.rank, hull, stern, bow, bridge),
+      favor: this.sumStat('favor', submarine.rank, hull, stern, bow, bridge)
+    };
+  }
+
+  public sumStat(statName: 'surveillance' | 'retrieval' | 'speed' | 'range' | 'favor', rank, ...parts) {
```

Comment:
rank and parts being implicit `any`, this method is a bit hard to understand, I'd prefer the signature to be self-explaining with proper typing 🙂